lib/pos_transaction_register.go

```diff
@@ -0,0 +1,655 @@
+	if aVal.minFeeNanosPerKB < bVal.minFeeNanosPerKB {
+		return 1
+	} else if aVal.minFeeNanosPerKB > bVal.minFeeNanosPerKB {
+		return -1
+	}
+
+	return 0
+}
+
+// AddTransaction adds a transaction to the register. If the transaction already exists in the register, or its size
+// exceeds the maximum mempool capacity, it is not added. Returns nil when transaction was successfully added to the
+// register, or an error otherwise.
+func (tr *TransactionRegister) AddTransaction(txn *MempoolTx) error {
+	tr.Lock()
+	defer tr.Unlock()
+
+	return tr.addTransactionNoLock(txn)
+}
+
+func (tr *TransactionRegister) addTransactionNoLock(txn *MempoolTx) error {
+	if txn == nil || txn.Hash == nil {
+		return fmt.Errorf("TransactionRegister.AddTransaction: Transaction or transaction hash is nil")
+	}
+
+	if _, ok := tr.txnMembership[*txn.Hash]; ok {
+		return nil
+	}
+
+	// If the transaction is too large, reject it.
+	if tr.totalTxnsSizeBytes > math.MaxUint64-txn.TxSizeBytes {
+		return fmt.Errorf("TransactionRegister.AddTransaction: Transaction size overflows uint64. Txn size %v, "+
+			"total size %v", txn.TxSizeBytes, tr.totalTxnsSizeBytes)
+	}
+
+	// Determine the min fee of the bucket based on the transaction's fee rate.
+	bucketMinFeeNanosPerKb, bucketMaxFeeNanosPerKB := computeFeeTimeBucketRangeFromFeeNanosPerKB(txn.FeePerKB,
+		tr.minimumNetworkFeeNanosPerKB, tr.feeBucketRateMultiplierBasisPoints)
+	// Lookup the bucket in the map.
+	bucket, bucketExists := tr.feeTimeBucketsByMinFeeMap[bucketMinFeeNanosPerKb]
+	if !bucketExists {
+		// If the bucket doesn't exist, create it and add the transaction to it.
+		bucket = NewFeeTimeBucket(bucketMinFeeNanosPerKb, bucketMaxFeeNanosPerKB)
+	}
+
+	// Add the transaction to the bucket.
+	if err := bucket.AddTransaction(txn); err != nil {
+		return errors.Wrapf(err, "TransactionRegister.AddTransaction: Error adding transaction to bucket: ")
+	}
+
+	if !bucketExists {
+		// If the bucket didn't exist, add it to the set and the map.
+		tr.addBucketNoLock(bucket)
+	}
+
+	tr.totalTxnsSizeBytes += txn.TxSizeBytes
+	tr.txnMembership[*txn.Hash] = txn
+	return nil
+}
+
+// RemoveTransaction removes a transaction from the register. If the transaction does not exist in the register, or its
+// size exceeds the current register size (which should never happen), it is not removed. Returns nil when transaction
+// was successfully removed from the register, or an error otherwise.
+func (tr *TransactionRegister) RemoveTransaction(txn *MempoolTx) error {
+	tr.Lock()
+	defer tr.Unlock()
+
+	return tr.removeTransactionNoLock(txn)
+}
+
+func (tr *TransactionRegister) removeTransactionNoLock(txn *MempoolTx) error {
+	if txn == nil || txn.Hash == nil {
+		return fmt.Errorf("TransactionRegister.RemoveTransaction: Transaction or transaction hash is nil")
+	}
+
+	if _, ok := tr.txnMembership[*txn.Hash]; !ok {
+		return nil
+	}
+
+	if tr.totalTxnsSizeBytes < txn.TxSizeBytes {
```

Comment:
A bit of an odd check. How could a transaction we're removing ever exceed the size of the whole transaction register? If it's a sanity-check, document that with a comment just so it's clear the reason why we're checking this.

Also if you want a sanity-check, what might be better is periodically checking that the sum of TxSizeBytes in tr.txnMembership is equal to tr.totalTxnsSizeBytes somehow. But not sure this is worth the performance hit, since you will have to lock it while you check this. Maybe if you architected a "once per minute" check or something.

Reply:
Ah yes this is a sanity check, comment added. It should never get triggered UNLESS somehow the referenced transaction changes. But we assume the MempoolTx type to be indempotent, so this sanity-check is redundant.

In terms of diagnostics, I think this would be handled by the struct that manages the mempool. I'm thinking a `mempool_manager.go` file would house this struct, and be responsible for giving other compoenents access to the mempool as well as doing things like diagnostics. The mempool_manager will run its own goroutine loop. 

I already started working on the architecture design enabling such mempool_manager struct to exist. You'll see it soon :)